pkg/ddl/partition.go

```diff
@@ -3869,54 +3837,66 @@ func (w *reorgPartitionWorker) BackfillData(handleRange reorgBackfillTask) (task
 		taskCtx.nextKey = nextKey
 		taskCtx.done = taskDone
 
-		isClustered := w.reorgedTbl.Meta().IsCommonHandle || w.reorgedTbl.Meta().PKIsHandle
-		if !isClustered {
-			// non-clustered table, we need to replace the _tidb_rowid handles since
-			// there may be duplicates across different partitions, due to EXCHANGE PARTITION.
-			// Meaning we need to check here if a record was double written to the new partition,
-			// i.e. concurrently written by StateWriteOnly or StateWriteReorganization.
-			// and we should skip it.
-			var found map[string][]byte
+		failpoint.InjectCall("PartitionBackfillData", len(w.rowRecords) > 0)
+		// For non-clustered tables, we need to replace the _tidb_rowid handles since
+		// there may be duplicates across different partitions, due to EXCHANGE PARTITION.
+		// Meaning we need to check here if a record was double written to the new partition,
+		// i.e. concurrently written by StateWriteOnly or StateWriteReorganization.
+		// and if so, skip it.
+		var found map[string][]byte
+		if len(w.oldKeys) > 0 {
+			// we must check if old IDs already been written,
+			// i.e. double written by StateWriteOnly or StateWriteReorganization.
+			// TODO: while waiting for BatchGet to check for duplicate, do another round of reads in parallel?
+			found, err = txn.BatchGet(ctx, w.oldKeys)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+
+		for i, prr := range w.rowRecords {
+			taskCtx.scanCount++
+			key := prr.key
+
 			if len(w.oldKeys) > 0 {
-				// we must check if old IDs already been written,
-				// i.e. double written by StateWriteOnly or StateWriteReorganization.
-				// The good thing is that we can then also skip the index generation for that row and we don't need to
-				// check if duplicate index entries was already copied either!
-				// TODO: while waiting for BatchGet to check for duplicate, do another round of reads in parallel?
-				found, err = txn.BatchGet(ctx, w.oldKeys)
+				if _, ok := found[string(w.oldKeys[i])]; ok {
```

Comment:
Does `len(w.oldKeys) > 0` mean it's a non clustered table? Maybe we could create a function for this if stmt and named it related with `non Clustered`. It's easy to understand.

Reply:
I will clarify with a comment, no need for a single use function I think.